### DISCUSSION TITLE
How to use pytest + subprocess

### DISCUSSION THREAD

**Discussion by AlexPaiva** [pyinstaller/pyinstaller#8090] [Help]
Hi,

So I am compiling to an .exe a multi threaded script that each thread calls a subprocess with 'pytest test.py', locally works perfectly obviously, but via pyinstaller it doesn't recognize the pytest command, even after all the data, binaries, etc additions and debugging, nothing.

Now I'm trying just plain pytest.main() calls but get issues with no stdin and stdout, I do mock stdin and stdout but then it only works for 1 pytest call, all others don't open at all and fail. 

Subprocess command I'm using is .run(): `subprocess.run(["pytest", "test.py"])`, I'm using the latest version of pyinstaller (6.x) and pytest, with python 3.

How can I make this work properly? Can anyone give me a hand with this?

PS: Goal is to run this on a machine without any python installation, meaning the .exe is fully independent

**Comment by rokm** [comment]
If your frozen application is calling `subprocess.run(["pytest", "test.py"])`, then you are implicitly assuming that your have an external python environment with `pytest` available. Which beats the point of freezing your code with PyInstaller in the first place.

**Reply by AlexPaiva** [reply]
How can I handle this? Freeze pytest and call it?

**Reply by rokm** [reply]
I don't know what you're really trying to achieve with that?

Yes, you should freeze `pytest`, but the invocation of subprocesses will then not be done via external `pytest` progam (which will still be unavailable), but via your frozen application's executable (`sys.executable`), and you will need to adjust your program flow accordingly (e.g., add a command-line option that denotes that the process is actually subprocess, and should launch the `pytest` codepath). Or, use `multiprocessing` (in that case, don't forget to call `multiprocessing.freeze_support`).

Even so, if you plan to run arbitrary .py scripts, be aware that you need to ensure that all their potential dependencies are collected beforehand in the program.

All in all, whatever you are trying to achieve here, I suspect that you shouldn't (unless you really know what you are doing).

**Reply by AlexPaiva** [reply]
Basically I use selenium-base for testing a website to show my directors, but to multi thread it the repo owner says only pytest works, so I made it work using subprocess and pytest calls, which works via command line and I can from my tkinter app call subprocesses of pytest without issues. (Each pytest process is the same function but it does randomized actions thus I need like 3 or 4 threads)

But I need this packaged in an .exe, everything works except this pytest subprocess thing. If I call it directly using their pytest.amin() caller I get issues with stdin and stdout and even after fixing them it only starts one thread do to maybe concurrency of stdin? The repo owner said do to isolation needs it has to be pytest else it doesn't work. How can I handle this? Replace `subprocess.run(["pytest", "test.py"])` with `subprocess.run([sys.executabl, "test.py"])` ?

**Reply by rokm** [reply]
Yes, but then you will need to handle this codepath in your main executable, and call `import pytest; pytest.main()` in the subprocess codepath (and handle running your test script the way calling `pytest test.py` would). 

See https://github.com/orgs/pyinstaller/discussions/5978#discussioncomment-971765 for a general example / idea of what we are talking about.

**Comment by bwoodsend** [comment]
Use `import pytest; pytest.main()` instead

**Reply by AlexPaiva** [reply]
Tried that, get issues with stdin and stdout, I mock it by transfering it to a file but this way it runs one thread and then all other threads don't start and the thread itself can't then be re-called or restarted, giving this error:
```
Traceback (most recent call last):
  File "main.py", line 521, in run_pytest_thread
  File "_pytest\config\__init__.py", line 150, in main
  File "_pytest\config\__init__.py", line 331, in _prepareconfig
  File "pluggy\_hooks.py", line 493, in __call__
  File "pluggy\_manager.py", line 115, in _hookexec
  File "pluggy\_callers.py", line 130, in _multicall
  File "_pytest\helpconfig.py", line 104, in pytest_cmdline_parse
  File "pluggy\_result.py", line 114, in get_result
  File "pluggy\_callers.py", line 77, in _multicall
  File "_pytest\config\__init__.py", line 1075, in pytest_cmdline_parse
  File "_pytest\config\__init__.py", line 1425, in parse
  File "_pytest\config\__init__.py", line 1327, in _preparse
  File "pluggy\_hooks.py", line 493, in __call__
  File "pluggy\_manager.py", line 115, in _hookexec
  File "pluggy\_callers.py", line 152, in _multicall
  File "pluggy\_result.py", line 114, in get_result
  File "pluggy\_callers.py", line 62, in _multicall
  File "_pytest\capture.py", line 149, in pytest_load_initial_conftests
  File "_pytest\capture.py", line 750, in start_global_capturing
  File "_pytest\capture.py", line 692, in _get_multicapture
  File "_pytest\capture.py", line 459, in __init__
OSError: [Errno 0] Error
```

Most likely an issue with stdin and stdout being mocked to a file, but even with monkeypatch do to concurrency there is no way to do this, easiest way is subprocess basically.

**Comment by bwoodsend** [comment]
To be clear `subprocess.run(["pytest", "test.py"])` is broken code even without PyInstaller. If you ran that from a Python environment whose `bin`/`Scripts` directory isn't in `PATH` then you'll either get a command not found error or possibly run from a different environment that is in `PATH`.

**Reply by AlexPaiva** [reply]
Without pyinstaller it works without any issues, see my previous replies to comments above to get a better understanding of my issue and my goal with pytest and pyinstaller

**Reply by bwoodsend** [reply]
It works only because your environment is set up in a certain way.

**Reply by AlexPaiva** [reply]
What would you recommend?

**Reply by rokm** [reply]
> What would you recommend?

To walk away and never look back.

**Reply by bwoodsend** [reply]
I'm with Rok. What you're attempting is already held together by spit and happenstance and you want to turn that into a standalone application with a framework that does not work with Python subprocesses....

**Reply by AlexPaiva** [reply]
If I wanted to run a python script via subprocess does pyinstaller support this? Like "python test.py"

**Reply by rokm** [reply]
> If I wanted to run a python script via subprocess does pyinstaller support this? Like "python test.py"

Then you would be running your script with *external* python interpreter, and you are implicitly assuming that you have one available in the environment. With all the dependencies required by `test.py`, at that. So there is little point in using PyInstaller in the first place (even for your GUI part).

If you want to run a script using interpreter that's bundled in PyInstaller-frozen application, see https://github.com/orgs/pyinstaller/discussions/5978#discussioncomment-971765. But since your `test.py` is not collected as a python module, nor is it the entry-point script for the frozen application (i.e., it is collected as a data file), its dependencies will not be automatically collected into the frozen application.

**Reply by AlexPaiva** [reply]
Made it work without pytest:
`subprocess.run([sys.executable, "--runi"])` and then on the main just add a if to check for the --runi argument and call a function from it and then 'exec(code)' on it, still looking into pytest tho.
